light-prover/prover/batch_append_circuit.go

```diff
@@ -0,0 +1,313 @@
+	hashChain := circuit.OldSubTreeHashChain
+	for i := 0; i < int(circuit.Depth); i++ {
+		hashChain = abstractor.Call(api, poseidon.Poseidon2{In1: hashChain, In2: circuit.Subtrees[i]})
+	}
+	return hashChain
+}
+
+func (circuit *BatchAppendCircuit) validateLeavesHashChain(api frontend.API) frontend.Variable {
+	hashChain := frontend.Variable(0)
+	for _, leaf := range circuit.Leaves {
+		hashChain = abstractor.Call(api, poseidon.Poseidon2{In1: hashChain, In2: leaf})
+	}
+	return hashChain
+}
+
+func (circuit *BatchAppendCircuit) batchAppend(api frontend.API) (frontend.Variable, []frontend.Variable) {
+	currentRoot := circuit.OldSubTreeHashChain
+	currentSubtrees := make([]frontend.Variable, len(circuit.Subtrees))
+	copy(currentSubtrees, circuit.Subtrees)
+
+	for i := 0; i < int(circuit.BatchSize); i++ {
+		index := api.Add(circuit.StartIndex, i)
+		leaf := circuit.Leaves[i]
+		currentRoot, currentSubtrees = circuit.append(api, leaf, currentSubtrees, index)
+	}
+
+	return currentRoot, currentSubtrees
+}
+
+func (circuit *BatchAppendCircuit) append(api frontend.API, leaf frontend.Variable, subtrees []frontend.Variable, index frontend.Variable) (frontend.Variable, []frontend.Variable) {
+	path := api.ToBinary(index, int(circuit.Depth))
+	currentNode := leaf
+
+	for i := 0; i < int(circuit.Depth); i++ {
+		isLeft := api.IsZero(path[i])
+		sibling := api.Select(isLeft, circuit.calculateZeroValue(api, i), subtrees[i])
+
+		left := api.Select(isLeft, currentNode, sibling)
+		right := api.Select(isLeft, sibling, currentNode)
+
+		currentNode = abstractor.Call(api, poseidon.Poseidon2{In1: left, In2: right})
+		subtrees[i] = api.Select(isLeft, currentNode, subtrees[i])
+	}
+
+	return currentNode, subtrees
+}
+
+func (circuit *BatchAppendCircuit) calculateZeroValue(api frontend.API, level int) frontend.Variable {
+	if level == 0 {
+		return 0
+	}
+	prevZero := circuit.calculateZeroValue(api, level-1)
+	return abstractor.Call(api, poseidon.Poseidon2{In1: prevZero, In2: prevZero})
+}
+
+func (circuit *BatchAppendCircuit) createSubtreesHashChain(api frontend.API, subtrees []frontend.Variable) frontend.Variable {
+	hashChain := frontend.Variable(0)
+	for _, subtree := range subtrees {
+		hashChain = abstractor.Call(api, poseidon.Poseidon2{In1: hashChain, In2: subtree})
+	}
+	return hashChain
+}
+
+type BatchAppendParameters struct {
+	OldSubTreeHashChain *big.Int   `json:"oldSubTreeHashChain"`
+	NewSubTreeHashChain *big.Int   `json:"newSubTreeHashChain"`
+	NewRoot             *big.Int   `json:"newRoot"`
+	HashchainHash       *big.Int   `json:"hashchainHash"`
+	StartIndex          uint32     `json:"startIndex"`
+	Leaves              []*big.Int `json:"leaves"`
+	Subtrees            []*big.Int `json:"subtrees"`
+	HashChainStartIndex uint32     `json:"hashChainStartIndex"`
+}
+
+func (p *BatchAppendParameters) TreeDepth() uint32 {
+	return 26
+}
+
+func (p *BatchAppendParameters) BatchSize() uint32 {
+	return uint32(len(p.Leaves))
+}
```

Comment:
note to be careful to keep the circuit general for any tree depth and batch size